File: phaseunwrapping.py
#!/usr/local/bin/python3

import numpy as np
from cv2 import *
import os
import sys


# Faire les directories:
wd=os.getcwd()
test=os.path.join(wd,"test/")
outputCapturePath=os.path.join(wd,"output/capture/")
outputPatternPath=os.path.join(wd,"output/pattern/")
outputWrappedPath=os.path.join(wd,"output/wrapped/")
outputUnwrappedPath=os.path.join(wd,"output/unwrapped/")
reliabilitiesPath=os.path.join(wd,"output/reliabilities/")
output_paths = [test,outputPatternPath, outputPatternPath , outputWrappedPath, outputUnwrappedPath, reliabilitiesPath]
for path in output_paths:
    if not os.path.exists(path):
        os.makedirs(path)


# La camera ----------------
CAM_OSNUMBER=1
# Dimensions du first monitor
monitor_width = int(2880); monitor_heigth = int(1800);
# Dimensions du first monitor
projector_width = int(1280) ; projector_heigth = int(720);

# Sinusoidal Patterns------------------------------------
# Set patterns parameters
params = structured_light_SinusoidalPattern_Params()
params.width=projector_width
params.height=projector_heigth
params.methodId=0
params.nbrOfPeriods=4;
params.setMarkers = True
params.horizontal = True
params.methodId = structured_light.PSP
params.shiftValue = 2.0 * np.pi / 3.0
params.nbrOfPixelsBetweenMarkers = 70
# Generate sinusoidal patterns
sinus=structured_light.SinusoidalPattern_create(params)
_, patternImages = sinus.generate()

# Enregistrer les patterns -----
for i in range(0, len(patternImages)):
    name = i + 1;
    imwrite(outputPatternPath + "pattern_{}".format(name) + ".png", patternImages[i]);
# ------------------------------

# MESURES --------------------------------------------------------------------
# A. Prendre une photo du damier sans frange
cap = VideoCapture(CAM_OSNUMBER)
if( not cap.isOpened() ):
    print( "Camera could not be opened")
    pass
ret, frame = cap.read()
retval=imwrite( test+"image_sans_frange" + ".png", frame);

# B. Prendre les photos de franges
# B.1. Afficher les franges et enregistrer les captures
namedWindow("pattern", WINDOW_NORMAL);
setWindowProperty("pattern", WND_PROP_FULLSCREEN, WINDOW_FULLSCREEN);
moveWindow("pattern", 0, 0);
imshow("pattern", patternImages);
waitKey(0)

nbrOfImages = 6;
count = 0;
captureImages=[]
while( count < nbrOfImages ):
    for i in range( 0, len(patternImages) ):
        imshow("pattern", patternImages[i]);
        waitKey(30)
        _, frame = cap.read()
        captureImages.append(frame)
        count += 1

# Fermer la caméra et le display window
cap.release()
moveWindow("pattern", 0, monitor_width);
destroyAllWindows()



greyImages=[]
for i in len(captureImages):
    greyImages.append( cvtColor(captureImages[i], COLOR_BGR2GRAY) )

#--------------------------------------------------------------------

# UNWRAPPING  -------------------------------------------------------
camSize = (captureImages[0].shape[0],captureImages[0].shape[1] )
paramsUnwrapping = phase_unwrapping_HistogramPhaseUnwrapping_Params()
paramsUnwrapping.height=projector_heigth;
paramsUnwrapping.width=projector_width;
phaseUnwrapping= phase_unwrapping_HistogramPhaseUnwrapping.create(paramsUnwrapping)



for  i in range( 1, nbrOfImages - 2):
    print('allo')

        if params.methodId == structured_light.FTP:

            for( i in range( 0, nbrOfImages) ):
                #We need three images to compute the shadow mask, as described in the reference paper
                # even if the phase map is computed from one pattern only
                captures=[]
                if( i == nbrOfImages - 2 ):
                    captures.append(greyImages[i]);
                    captures.append(greyImages[i-1]);
                    captures.append(greyImages[i+1]);
                else if( i == nbrOfImages - 1 ):
                    captures.push_back(greyImages[i]);
                    captures.push_back(greyImages[i-1]);
                    captures.push_back(greyImages[i-2]);
                else:
                    captures.push_back(greyImages[i]);
                    captures.push_back(greyImages[i+1]);
                    captures.push_back(greyImages[i+2]);

            sinus.computePhaseMap(captureImages)
            sinus.unwrapPhaseMap(wrappedPhaseMap, unwrappedPhaseMap, camSize, shadowMask);

                    phaseUnwrappingunwrapPhaseMap(wrappedPhaseMap, unwrappedPhaseMap, shadowMask);
                    Mat reliabilities, reliabilities8;
                    phaseUnwrapping->getInverseReliabilityMap(reliabilities);
                    reliabilities.convertTo(reliabilities8, CV_8U, 255,128);
                    ostringstream tt;
                    tt << i;
                    imwrite(reliabilitiesPath + "rel_{}".format(i) + ".png", reliabilities8);

                    if( !outputUnwrappedPhasePath.empty() )
                    {
                        ostringstream name;
                        name << i;
                        imwrite(outputUnwrappedPhasePath + "_FTP_" + name.str() + ".png", unwrappedPhaseMap8);
                    }
                    if( !outputWrappedPhasePath.empty() )
                    {
                        ostringstream name;
                        name << i;
                        imwrite(outputWrappedPhasePath + "_FTP_" + name.str() + ".png", wrappedPhaseMap8);
                    }
                }
                break;

    else:

        captures=[];
        for j in range(0,3):
            captures.append(greyImages[i+j]);

        # Wrap/Unwrap ------------
        wrappedPhaseMap, shadowMask = sinus.computePhaseMap(captures)
        print("wrappedPhaseMap succes!")
        wrappedPhaseMapScaled = convertScaleAbs(wrappedPhaseMap, 1.0, 128)
        wrappedPhaseMap8 = wrappedPhaseMapScaled.astype(np.uint8)
        shadowMaskScaled = convertScaleAbs(shadowMask, 1.0, 128)
        shadowMask8 = shadowMaskScaled.astype(np.uint8)
        # ----------------------

        # Enregistrement ------------
        # Wrapped Phase Map
        if( params.methodId == structured_light.PSP ):
            imwrite(outputWrappedPath + "_PSP_{}".format(i) + ".png", wrappedPhaseMap8);
        else:
            imwrite(outputWrappedPath + "_FAPS_{}".format(i)  + ".png", wrappedPhaseMap8);
        # ShadowMask
        if( params.methodId == structured_light.PSP ):
            imwrite(outputWrappedPath + "sm_PSP_{}".format(i) + ".png", shadowMask8);
        else:
            imwrite(outputWrappedPath + "sm_FAPS_{}".format(i)  + ".png", shadowMask8);
        # Capture images
        if( params.methodId == structured_light.PSP ):
            imwrite(outputCapturePath + "_PSP_{}".format(i)+ ".png", captureImages[i]);
        else:
            imwrite(outputCapturePath + "_FAPS_{}".format(i) + ".png", captureImages[i]);
        if( i == nbrOfImages - 3 ):
            if( params.methodId == structured_light.PSP ):
                nameBis = i+1;
                nameTer = i+2;
                imwrite(outputCapturePath + "_PSP_{}".format(nameBis) + ".png", captureImages[i+1]);
                imwrite(outputCapturePath + "_PSP_{}".format(nameTer) + ".png", captureImages[i+2]);
            else:
                nameBis = i+1;
                nameTer = i+2;
                imwrite(outputCapturePath + "_FAPS_{}".format(nameBis)+ ".png", captureImages[i+1]);
                imwrite(outputCapturePath + "_FAPS_{}".format(nameTer) + ".png", captureImages[i+2]);
        print("saved")
        # ----------------------

        # Unwrap ---------------
        sinus.unwrapPhaseMap(wrappedPhaseMap, camSize, shadowMask)
        print("unwrappedPhaseMap succes!")
        unwrappedPhaseMapScaled = convertScaleAbs(unwrappedPhaseMap, 255, 128)
        unwrappedPhaseMap8 = unwrappedPhaseMapScaled.astype(np.uint8)
        # ----------------------

        # Enregistrement ------------
        # Unwrapped Phase Map
        if( params.methodId == structured_light.PSP ):
            imwrite(outputUnwrappedPath + "_PSP_{}".format(i) + ".png", unwrappedPhaseMap8);
        else:
            imwrite(outputUnwrappedPath + "_FAPS_{}".format(i)  + ".png", unwrappedPhaseMap8);
        # --------------------------


print("Done")
--------------------------------------------------------------------

File: pu-opencv/phaseunwrapping.cpp
#include <opencv2/highgui.hpp>
#include <vector>
#include <iostream>
#include <fstream>
#include <opencv2/core.hpp>
#include <opencv2/core/utility.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/calib3d.hpp>
#include <opencv2/structured_light.hpp>
#include <opencv2/phase_unwrapping.hpp>
using namespace cv;
using namespace std;
static const char* keys =
{
    "{@width | | Projector width}"
    "{@height | | Projector height}"
    "{@periods | | Number of periods}"
    "{@setMarkers | | Patterns with or without markers}"
    "{@horizontal | | Patterns are horizontal}"
    "{@methodId | | Method to be used}"
    "{@outputPatternPath | | Path to save patterns}"
    "{@outputWrappedPhasePath | | Path to save wrapped phase map}"
    "{@outputUnwrappedPhasePath | | Path to save unwrapped phase map}"
    "{@outputCapturePath | | Path to save the captures}"
    "{@reliabilitiesPath | | Path to save reliabilities}"
};
static void help()
{
    cout << "\nThis example generates sinusoidal patterns" << endl;
    cout << "To call: ./example_structured_light_createsinuspattern <width> <height>"
            " <number_of_period> <set_marker>(bool) <horizontal_patterns>(bool) <method_id>"
            " <output_captures_path> <output_pattern_path>(optional) <output_wrapped_phase_path> (optional)"
            " <output_unwrapped_phase_path>" << endl;
}
int main(int argc, char **argv)
{
    if( argc < 2 )
    {
        help();
        return -1;
    }
    structured_light::SinusoidalPattern::Params params;
    phase_unwrapping::HistogramPhaseUnwrapping::Params paramsUnwrapping;
    // Retrieve parameters written in the command line
    CommandLineParser parser(argc, argv, keys);
    params.width = parser.get<int>(0);
    params.height = parser.get<int>(1);
    params.nbrOfPeriods = parser.get<int>(2);
    params.setMarkers = parser.get<bool>(3);
    params.horizontal = parser.get<bool>(4);
    params.methodId = parser.get<int>(5);
    String outputCapturePath = parser.get<String>(6);
    params.shiftValue = static_cast<float>(2 * CV_PI / 3);
    params.nbrOfPixelsBetweenMarkers = 70;
    String outputPatternPath = parser.get<String>(7);
    String outputWrappedPhasePath = parser.get<String>(8);
    String outputUnwrappedPhasePath = parser.get<String>(9);
    String reliabilitiesPath = parser.get<String>(10);
    Ptr<structured_light::SinusoidalPattern> sinus =
            structured_light::SinusoidalPattern::create(makePtr<structured_light::SinusoidalPattern::Params>(params));
    Ptr<phase_unwrapping::HistogramPhaseUnwrapping> phaseUnwrapping;
    vector<Mat> patterns;
    Mat shadowMask;
    Mat unwrappedPhaseMap, unwrappedPhaseMap8;
    Mat wrappedPhaseMap, wrappedPhaseMap8;
    //Generate sinusoidal patterns
    sinus->generate(patterns);


    VideoCapture cap(1);
    if( !cap.isOpened() )
    {
        cout << "Camera could not be opened" << endl;
        return -1;
    }
    cap.set(CAP_PROP_PVAPI_PIXELFORMAT, CAP_PVAPI_PIXELFORMAT_MONO8);
    namedWindow("pattern", WINDOW_NORMAL);
    setWindowProperty("pattern", WND_PROP_FULLSCREEN, WINDOW_FULLSCREEN);
    imshow("pattern", patterns[0]);
    cout << "Press any key when ready" << endl;
    waitKey(0);
    int nbrOfImages = 30;
    int count = 0;
    vector<Mat> img(nbrOfImages);
    Size camSize(-1, -1);
    while( count < nbrOfImages )
    {
        for(int i = 0; i < (int)patterns.size(); ++i )
        {
            Mat colorimg;
            imshow("pattern", patterns[i]);
            waitKey(300);
            cap >> colorimg;
            cvtColor(colorimg, img[count], COLOR_BGR2GRAY);
            count += 1;
        }
    }
    destroyWindow("pattern");
    cout << "press enter when ready" << endl;

    switch(params.methodId)
    {
        case structured_light::FTP:
            for( int i = 0; i < nbrOfImages; ++i )
            {
                /*We need three images to compute the shadow mask, as described in the reference paper
                 * even if the phase map is computed from one pattern only
                */
                vector<Mat> captures;
                if( i == nbrOfImages - 2 )
                {
                    captures.push_back(img[i]);
                    captures.push_back(img[i-1]);
                    captures.push_back(img[i+1]);
                }
                else if( i == nbrOfImages - 1 )
                {
                    captures.push_back(img[i]);
                    captures.push_back(img[i-1]);
                    captures.push_back(img[i-2]);
                }
                else
                {
                    captures.push_back(img[i]);
                    captures.push_back(img[i+1]);
                    captures.push_back(img[i+2]);
                }
                sinus->computePhaseMap(captures, wrappedPhaseMap, shadowMask);
                if( camSize.height == -1 )
                {
                    camSize.height = img[i].rows;
                    camSize.width = img[i].cols;
                    paramsUnwrapping.height = camSize.height;
                    paramsUnwrapping.width = camSize.width;
                    phaseUnwrapping =
                    phase_unwrapping::HistogramPhaseUnwrapping::create(paramsUnwrapping);
                }
                sinus->unwrapPhaseMap(wrappedPhaseMap, unwrappedPhaseMap, camSize, shadowMask);
                phaseUnwrapping->unwrapPhaseMap(wrappedPhaseMap, unwrappedPhaseMap, shadowMask);
                Mat reliabilities, reliabilities8;
                phaseUnwrapping->getInverseReliabilityMap(reliabilities);
                reliabilities.convertTo(reliabilities8, CV_8U, 255,128);
                ostringstream tt;
                tt << i;
                imwrite(reliabilitiesPath + tt.str() + ".png", reliabilities8);
                unwrappedPhaseMap.convertTo(unwrappedPhaseMap8, CV_8U, 1, 128);
                wrappedPhaseMap.convertTo(wrappedPhaseMap8, CV_8U, 255, 128);
                if( !outputUnwrappedPhasePath.empty() )
                {
                    ostringstream name;
                    name << i;
                    imwrite(outputUnwrappedPhasePath + "_FTP_" + name.str() + ".png", unwrappedPhaseMap8);
                }
                if( !outputWrappedPhasePath.empty() )
                {
                    ostringstream name;
                    name << i;
                    imwrite(outputWrappedPhasePath + "_FTP_" + name.str() + ".png", wrappedPhaseMap8);
                }
            }
            break;
        case structured_light::PSP:
        case structured_light::FAPS:
            for( int i = 0; i < nbrOfImages - 2; ++i )
            {
                vector<Mat> captures;
                captures.push_back(img[i]);
                captures.push_back(img[i+1]);
                captures.push_back(img[i+2]);
                sinus->computePhaseMap(captures, wrappedPhaseMap, shadowMask);
                if( camSize.height == -1 )
                {
                    camSize.height = img[i].rows;
                    camSize.width = img[i].cols;
                    paramsUnwrapping.height = camSize.height;
                    paramsUnwrapping.width = camSize.width;
                    phaseUnwrapping =
                    phase_unwrapping::HistogramPhaseUnwrapping::create(paramsUnwrapping);
                }
                sinus->unwrapPhaseMap(wrappedPhaseMap, unwrappedPhaseMap, camSize, shadowMask);
                unwrappedPhaseMap.convertTo(unwrappedPhaseMap8, CV_8U, 1, 128);
                wrappedPhaseMap.convertTo(wrappedPhaseMap8, CV_8U, 255, 128);
                phaseUnwrapping->unwrapPhaseMap(wrappedPhaseMap, unwrappedPhaseMap, shadowMask);
                Mat reliabilities, reliabilities8;
                phaseUnwrapping->getInverseReliabilityMap(reliabilities);
                reliabilities.convertTo(reliabilities8, CV_8U, 255,128);
                ostringstream tt;
                tt << i;
                imwrite(reliabilitiesPath + tt.str() + ".png", reliabilities8);
                if( !outputUnwrappedPhasePath.empty() )
                {
                    ostringstream name;
                    name << i;
                    if( params.methodId == structured_light::PSP )
                        imwrite(outputUnwrappedPhasePath + "_PSP_" + name.str() + ".png", unwrappedPhaseMap8);
                    else
                        imwrite(outputUnwrappedPhasePath + "_FAPS_" + name.str() + ".png", unwrappedPhaseMap8);
                }
                if( !outputWrappedPhasePath.empty() )
                {
                    ostringstream name;
                    name << i;
                    if( params.methodId == structured_light::PSP )
                        imwrite(outputWrappedPhasePath + "_PSP_" + name.str() + ".png", wrappedPhaseMap8);
                    else
                        imwrite(outputWrappedPhasePath + "_FAPS_" + name.str() + ".png", wrappedPhaseMap8);
                }
                if( !outputCapturePath.empty() )
                {
                    ostringstream name;
                    name << i;
                    if( params.methodId == structured_light::PSP )
                        imwrite(outputCapturePath + "_PSP_" + name.str() + ".png", img[i]);
                    else
                        imwrite(outputCapturePath + "_FAPS_" + name.str() + ".png", img[i]);
                    if( i == nbrOfImages - 3 )
                    {
                        if( params.methodId == structured_light::PSP )
                        {
                            ostringstream nameBis;
                            nameBis << i+1;
                            ostringstream nameTer;
                            nameTer << i+2;
                            imwrite(outputCapturePath + "_PSP_" + nameBis.str() + ".png", img[i+1]);
                            imwrite(outputCapturePath + "_PSP_" + nameTer.str() + ".png", img[i+2]);
                        }
                        else
                        {
                            ostringstream nameBis;
                            nameBis << i+1;
                            ostringstream nameTer;
                            nameTer << i+2;
                            imwrite(outputCapturePath + "_FAPS_" + nameBis.str() + ".png", img[i+1]);
                            imwrite(outputCapturePath + "_FAPS_" + nameTer.str() + ".png", img[i+2]);
                        }
                    }
                }
            }
            break;
        default:
            cout << "error" << endl;
    }
    cout << "done" << endl;
    if( !outputPatternPath.empty() )
    {
        for( int i = 0; i < 3; ++ i )
        {
            ostringstream name;
            name << i + 1;
            imwrite(outputPatternPath + name.str() + ".png", patterns[i]);
        }
    }

    return 0;
}
